src/connectors/impls/gpubsub/gsub.rs

```diff
@@ -0,0 +1,279 @@
+
+                        // If we can't send errors to the main task, disconnect and let it restart
+                        return;
+                    }
+
+                    // There was an error other than timeout, let the main task handle the failure, but
+                    // do not drop the connection.
+                    continue;
+                }
+            },
+            Err(timeout_error) => {
+                if let Err(e) = sender.send(Err(timeout_error.into())).await {
+                    error!(
+                        "Failed to send a PubSub timeout error to the main task: {}",
+                        e
+                    );
+
+                    // If we can't send errors to the main task, disconnect and let it restart
+                    return;
+                }
+
+                // If we have a timeout, exit the task and let the main task handle reconnection
+                return;
+            }
+        };
+
+        for message in response.received_messages {
+            let ReceivedMessage {
+                ack_id,
+                message: msg,
+                ..
+            } = message;
+
+            if let Some(PubsubMessage { data, .. }) = msg {
+                ack_counter += 1;
+                ack_ids.insert(ack_counter, ack_id);
+                if let Err(e) = sender.send(Ok((ack_counter, data))).await {
+                    error!("Failed to send a PubSub message to the main task: {}", e);
+
+                    // If we can't send to the main task, disconnect and let it restart
+                    return;
+                }
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Source for GSubSource {
+    async fn connect(&mut self, _ctx: &SourceContext, _attempt: &Attempt) -> Result<bool> {
+        let token = Token::new()?;
+
+        let tls_config = ClientTlsConfig::new()
+            .ca_certificate(Certificate::from_pem(googapis::CERTIFICATES))
+            .domain_name("pubsub.googleapis.com");
+
+        let channel = Channel::from_static("https://pubsub.googleapis.com")
+            .connect_timeout(Duration::from_nanos(self.config.connect_timeout))
+            .tls_config(tls_config)?
+            .connect()
+            .await;
+
+        let channel = dbg!(channel)?;
+
+        let client = SubscriberClient::with_interceptor(
+            channel,
+            AuthInterceptor {
+                token: Box::new(move || match token.header_value() {
+                    Ok(val) => Ok(val),
+                    Err(_) => Err(Status::unavailable(
+                        "Failed to retrieve authentication token.",
+                    )),
+                }),
+            },
+        );
+
+        let client = Arc::new(Mutex::new(client));
+
+        let (tx, rx) = async_std::channel::bounded(QSIZE.load(Ordering::Relaxed));
+
+        async_std::task::spawn(consumer_task(
```

Comment:
keep track of the handle and delete any running one at the start of this function. Otherwise you end up having several consuming tasks running in concurrently.